src/Libraries/RevitNodes/GeometryConversion/ProtoToRevitFace.cs

```diff
@@ -0,0 +1,154 @@
+       {
+           try
+           {
+               return ToRevitType(sol, performHostUnitConversion, BRepType.Solid);
+           }
+           catch
+           {
+               return null;
+           }
+       }
+
+       /// <summary>
+       /// this method attmempts to construct a BRep from a surface.
+       /// </summary>
+       /// <param name="surf"></param>
+       /// <param name="performHostUnitConversion"></param>
+       /// <returns></returns>
+       public static GeometryObject ToRevitType(Surface surf,
+          bool performHostUnitConversion = true)
+       {
+           try
+           {
+               return ToRevitType(surf, performHostUnitConversion, BRepType.OpenShell);
+           }
+           catch
+           {
+               return null;
+           }
+       }
+
+       private static GeometryObject ToRevitType(Autodesk.DesignScript.Geometry.Topology topology,
+       bool performHostUnitConversion,
+       BRepType type)
+       {
+           List<Face> faces = null;
+
+           if (performHostUnitConversion)
+           {
+               var newTopology = topology.InHostUnits();
+               faces = newTopology.Faces.ToList();
+               newTopology.Dispose();
+           }
+           else
+           {
+               faces = topology.Faces.ToList();
+           }
+
+           var brb = new BRepBuilder(type);
+           var edgeDict = new Dictionary<Edge, BRepBuilderGeometryId>();
+
+           //foreach face in solid
+           foreach (Face protoFace in faces)
+           {
+               var geom = protoFace.SurfaceGeometry();
+
+               var ngeom = geom.ToNurbsSurface();
+               bool flipped = false;
+
+               // Check if the nurbs surface has flipped compared to the original surface
+               if (geom.NormalAtParameter(.5, .5).Dot(ngeom.NormalAtParameter(.5, .5)) < 0)
+               {
+                   flipped = true;
+               }
+
+               // Create Revit nurbs surface
+               var bbface = BRepBuilderSurfaceGeometry.CreateNURBSSurface(ngeom.DegreeU, ngeom.DegreeV,
+                   ngeom.UKnots(), ngeom.VKnots(), ngeom.ControlPoints().SelectMany(x => x.Select(y => y.ToXyz())).ToList(),
+                   ngeom.Weights().SelectMany(x => x).ToList(),
+                   false,
+                   new BoundingBoxUV());
+
+               // Dispose geometries
+               geom.Dispose();
+               ngeom.Dispose();
+
+               // Add face
+               var faceId = brb.AddFace(bbface, flipped);
+
+               // add loops and connected edges
+               foreach (var loop in protoFace.Loops)
+               {
+                   var loopId = brb.AddLoop(faceId);
+
+                   foreach (var coedge in loop.CoEdges)
+                   {
+                       var edge = coedge.Edge;
+                       BRepBuilderGeometryId edgeId;
+                       if (edgeDict.ContainsKey(edge))
```

Comment:
I'm a little concerned here, if it this dictionary lookup seems to be working thats great! But how is hashcode implemented for edges? I'm more curious than anything because I know this didn't used to work :smile: